File: #55_Jump_Game/Solution.java
/* greedy approach
 * keep track of the farthest index the array can reach */

public class Solution {
    public boolean canJump(int[] nums) {
        if (nums == null || nums.length == 0) return false;
        if (nums.length == 1) return true;
        int maxIndex = 0;
        for (int i = 0; i < nums.length - 1; i++) {
        	if (i > maxIndex) break;
        	maxIndex = Math.max(maxIndex, nums[i] + i);
        }
        return maxIndex >= nums.length - 1;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	int[] t = {2, 3, 1, 1, 4};
    	System.out.println(sol.canJump(t));
    }
}

File: #55_Jump_Game/Solution.py
'''
Greedy, O(n) Solution
'''

class Solution:
	# @param {integer[]} nums
	# @return {boolean}
	def canJump(self, nums):
		if not nums:
			return True
		maxIndex, cur = 0, 0
		while cur < len(nums) and cur <= maxIndex:
			maxIndex = max(nums[cur] + cur, maxIndex)
			cur += 1
		return cur == len(nums)

sol = Solution()
print sol.canJump([2,3,1,1,4])